### PR TITLE
Create bug-1690385-pref-activity-stream-new-tab-default-content-relea…

### PR DESCRIPTION
**Pull request by SuYoungHong**
…se-85-87

jetstream config file for new-tab-default-content-release-85-87

@tdsmith , can you review?

### CODE REVIEW

File: bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87 [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+[segments.attributed]
+select_expression = 'attr_source is not null'
+data_source = "npp"
+
+[segments.non_attributed]
+select_expression = 'attr_source is null'
+data_source = "npp"
+
+[segments.data_sources.npp]
```
Comment (tdsmith) [inline bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87:36]: When are you expecting to receive a new_profile ping from these clients? If you think you'll always receive them on the same day as the enrollment ping, then this will work the way you're expecting. If you think you might have received a new_profile ping on a day _before_ the client enrolls, then we'll need to provide a `window_start` value (if the profiles might be just a day or two old) or do some aggregation aforethought (if profiles might be months old).

File: bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87 [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+## Segments
+
+[segments]
+
+[segments.attributed]
+select_expression = 'attr_source is not null'
+data_source = "npp"
+
+[segments.non_attributed]
+select_expression = 'attr_source is null'
```
Comment (tdsmith) [inline bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87:33]: fwiw this segment will include clients where the distribution_id is not null, since they will not appear in the segment table, so those clients will have null values in all columns provided by the segment table after the join onto the enrollments -- not sure if that's what you want

File: bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87 [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+[metrics.homepage_search_count.statistics.bootstrap_mean]
+
+
+## topsite_clicks
+
+[metrics.topsite_clicks]
+select_expression = "COUNTIF(
+                event = 'CLICK'
+                AND source = 'TOP_SITES')"
```
Comment (tdsmith) [inline bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87:74]: you can't have newlines in double-quoted strings unless you backslash-escape them; https://toml.io/en/v1.0.0#string

Try triple-quotes, just like Python:

```toml
select_expression = """COUNTIF(
                event = 'CLICK'
                AND source = 'TOP_SITES')"""
```

here and below

File: bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87 [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+[metrics.highlight_clicks.statistics.bootstrap_mean]
+
+
+## homepage_switch
+    # NOTE: this metric doesn't take into account users setting their page to 
+    # blank. Only when users have set the page to a 3rd party! 
+
+[metrics.homepage_switch]
+select_expression = "MAX(
```
Comment (tdsmith) [inline bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87:96]: If a client never sends an activity stream event ping this will evaluate to NULL and the stat will choke. You should either a) write something like COALESCE(MAX(...), FALSE) to provide a default value, or use `pre_treatments = ["drop_nulls"]` on the statistic declaration if it doesn't make sense to include those users in the calculation in any way.

**Review by tdsmith — COMMENTED** [review]
The file needs to have a .toml extension to be interpreted as a configuration and validated, so that'll need to be fixed.

The metrics here make sense to me!

fwiw we also have friendly_name and description fields that you can define on metrics; they aren't displayed on the result pages _yet_ but they _will_ be. There are examples in https://github.com/mozilla/jetstream-config/blob/main/custom-messaging-in-aboutwelcome-for-chrome-users-to-import.toml

File: bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87 [hunk 5]
```diff
@@ -0,0 +1,152 @@
+[experiment]
+end_date = '2020-04-13'
```
[the remark this answers is not in the document]
Comment (SuYoungHong) [reply bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87:2]: +1

File: bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87 [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+segments = ["attributed", "non_attributed"]
+
+enrollment_period = 8
+
+reference_branch = "baseline-newtab-content"
```
[the remark this answers is not in the document]
Comment (SuYoungHong) [reply bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87:8]: hmm, what do you mean by data in Experimenter will be wrong? Specifically for reference_branch. 

I see that there's a enrollment period parameter in experimenter, but I'm not seeing any obvious way where we designate a control or reference branch in experimenter.

File: bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87 [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+
+## data sources
+
+[data_sources]
```
[the remark this answers is not in the document]
Comment (SuYoungHong) [reply bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87:13]: +1

File: bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87 [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+[segments.data_sources.npp]
+from_expression = """
+    (SELECT
+        DATE(submission_timestamp) AS submission_date,
+        client_id,
+        environment.settings.attribution.source AS attr_source
+    FROM `moz-fx-data-shared-prod.telemetry.new_profile`
+    WHERE DATE(submission_timestamp) BETWEEN "{{experiment.start_date_str}}" AND "{{experiment.last_enrollment_date_str}}"
+    AND environment.partner.distribution_id IS NULL
```
[the remark this answers is not in the document]
Comment (SuYoungHong) [reply bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87:44]: +1

File: bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87 (hunk 1 above)
Comment (SuYoungHong) [reply bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87:36]: got it. will assume majority of users report same day for simplicity right now, but will keep an eye out if it's worth windowing them for completeness.

File: bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87 (hunk 2 above)
Comment (SuYoungHong) [reply bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87:33]: ahh that's good to know. 

that is, indeed my intention:
* attributed = non-null attribution source
* non-attributed = we have NPP without attribution + we have NPP but distribution + we don't have NPP

So it sounds like that select expression logic is executing AFTER we've joined the enrollments with the segements data source table, is that correct?

File: bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87 (hunk 1 above)
Comment (SuYoungHong) [reply bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87:36]: hmm, actually... considering my below comment, I think I actually want to be conservative here, and use the `window_start` value here... to not mislabel any users. 

is there any docs on how to specify a window start for a segements data source? I didn't catch it in the docs.

File: bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87 (hunk 3 above)
Comment (SuYoungHong) [reply bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87:74]: good catch, +1

File: bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87 (hunk 4 above)
Comment (SuYoungHong) [reply bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87:96]: ah, I see. I do want to provide a default value here. 

So I'm guessing the select expression is evaluating after the enrollment table has been joined to the data source, correct? 

+1 on the heads up. 

Hot take, the COALESCE solution works, but feels a bit like a hack-y (and requires a bit of mozanalysis plumbing context). Would be good to call this out as a common use case, or maybe parameterize it like we pre_treatments. 

Specificially, I mean, "how do you want to deal with values when there's isn't a match on the join".

File: bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87 (hunk 6 above)
[the remark this answers is not in the document]
Comment (tdsmith) [reply bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87]: I guess this is a Normandy experiment, yeah? I think the branch at the top of the page gets the flag; it's more obvious in the "edit branches" screen.

File: bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87 (hunk 1 above)
Comment (tdsmith) [reply bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87]: The clearest explanation is https://mozilla.github.io/mozanalysis/api/segments.html#mozanalysis.segments.SegmentDataSource -- Jetstream accepts window_start and window_end arguments. It'd look like e.g. `window_start = -1`.

File: bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87 (hunk 2 above)
Comment (tdsmith) [reply bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87]: Yes, that's a good model.

File: bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87 (hunk 4 above)
Comment (tdsmith) [reply bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87]: The premise, which I think is basically right, is that the user must always make a decision about how to handle NULLs. Right now the metrics definitions are a thin wrapper around mozanalysis metrics, which spit the select_expression directly into a query.

An alternative and maybe more elegant user interface could be something like adding a mandatory `if_null` key to every metric definition that accepts either a value or e.g. `"drop"`. That at least brings "you forgot to handle nulls" errors forward to the point of definition, which I like a lot.

But the Jetstream Metric object will have to diverge a little from the mozanalysis Metric in that case because mozanalysis can't actually drop nulls in the results dataframes that it emits, because there's one row per _client_ and we definitely don't want to remove that client from other metrics just because they had a null for one of them. That's why we've associated the drop_nulls pretreatment with statistics, which operate columnwise on a single metric.

And because the Jetstream metrics diverge from the mozanalysis metrics, defining a metric in mozanalysis will no longer be sufficient for a metric to be usable with Jetstream, because we need to supply the additional context somewhere about how to handle nulls. This is already sort of true because we have to provide default statistics for them. Maybe mozanalysis metrics get a special warrant that the null handling is correct, idk.

tl;dr there are some tradeoffs, but I agree this is clunky, so maybe they're worth it.

**Comment by tdsmith** [comment]
The error you're seeing in CI (`ValueError: Please define a statistical treatment for the metric pocket_rec_clicks`) is actually a good an example of the point that metrics defined in mozanalysis aren't usable out of the box anyway without a little more information.

Adding something like

```toml
[metrics.pocket_rec_clicks.statistics.bootstrap_mean]
[metrics.pocket_rec_clicks.statistics.deciles]
```

(and also for pocket_spoc_clicks) should move this towards green.

Incidentally, `'days_of_use', 'search_count', 'organic_search_count'` all come for free as overall metrics in the [default config](https://github.com/mozilla/jetstream/blob/main/jetstream/config/default_metrics.toml) so specifying them again doesn't do anything -- though it doesn't hurt, either.

File: bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87 (hunk 6 above)
[the remark this answers is not in the document]
Comment (SuYoungHong) [reply bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87]: +1

File: bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87 (hunk 1 above)
Comment (SuYoungHong) [reply bug-1690385-pref-activity-stream-new-tab-default-content-release-85-87]: sounds good. 

I added the `window_start` params into the `[segments.data_sources.npp]` data source section, based on the documentation (it seems to be defined where you define the data source). 

Please correct me if I'm wrong.

**Comment by SuYoungHong** [comment]
@tdsmith , hmm, the tests are still failing. I think the segmenting code might have a bug in it (I suspect it's inside mozanalysis).

**Comment by tdsmith** [comment]
Maybe a docs bug :) The code's fine. https://github.com/mozilla/jetstream-config#defining-segments illustrates out that the segment definition should include an aggregation function. Lines 29-52 are the subquery for the segment labels; you'll notice that it's grouped by client_id and branch.

So you'll want to define the segments like `LOGICAL_OR(attr_source) IS NULL` (or `IS NOT NULL`).

**Comment by SuYoungHong** [comment]
oh I see! you're right, my mistake!
